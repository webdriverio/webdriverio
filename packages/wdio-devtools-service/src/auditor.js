import Diagnostics from 'lighthouse/lighthouse-core/audits/diagnostics'
import MainThreadWorkBreakdown from 'lighthouse/lighthouse-core/audits/mainthread-work-breakdown'
import Metrics from 'lighthouse/lighthouse-core/audits/metrics'
import ServerResponseTime from 'lighthouse/lighthouse-core/audits/server-response-time'
import CumulativeLayoutShift from 'lighthouse/lighthouse-core/audits/metrics/cumulative-layout-shift'
import FirstContentfulPaint from 'lighthouse/lighthouse-core/audits/metrics/first-contentful-paint'
import LargestContentfulPaint from 'lighthouse/lighthouse-core/audits/metrics/largest-contentful-paint'
import SpeedIndex from 'lighthouse/lighthouse-core/audits/metrics/speed-index'
import InteractiveMetric from 'lighthouse/lighthouse-core/audits/metrics/interactive'
import TotalBlockingTime from 'lighthouse/lighthouse-core/audits/metrics/total-blocking-time'
import ReportScoring from 'lighthouse/lighthouse-core/scoring'
import defaultConfig from 'lighthouse/lighthouse-core/config/default-config'
import logger from '@wdio/logger'

const log = logger('@wdio/devtools-service:Auditor')

const SHARED_AUDIT_CONTEXT = {
    settings: { throttlingMethod: 'devtools' },
    LighthouseRunWarnings: false,
    computedCache: new Map()
}

export default class Auditor {
    constructor (traceLogs, devtoolsLogs) {
        this.devtoolsLogs = devtoolsLogs
        this.traceLogs = traceLogs
        this.url = traceLogs.pageUrl
        this.loaderId = traceLogs.loaderId
    }

    _audit (AUDIT, params = {}) {
        const auditContext = {
            options: { ...AUDIT.defaultOptions },
            ...SHARED_AUDIT_CONTEXT
        }

        try {
            return AUDIT.audit({
                traces: { defaultPass: this.traceLogs },
                devtoolsLogs: { defaultPass: this.devtoolsLogs },
                TestedAsMobileDevice: true,
                ...params
            }, auditContext)
        } catch (e) {
            log.error(e)
            return {}
        }
    }

    /**
     * an Auditor instance is created for every trace so provide an updateCommands
     * function to receive the latest performance metrics with the browser instance
     */
    updateCommands (browser, customFn) {
        const commands = Object.getOwnPropertyNames(Object.getPrototypeOf(this)).filter(
            fnName => fnName !== 'constructor' && fnName !== 'updateCommands' && !fnName.startsWith('_'))
        commands.forEach(fnName => browser.addCommand(fnName, customFn || ::this[fnName]))
    }

    async getMainThreadWorkBreakdown () {
        const result = await this._audit(MainThreadWorkBreakdown)
        return result.details.items.map(
            ({ group, duration }) => ({ group, duration })
        )
    }

    async getDiagnostics () {
        const result = await this._audit(Diagnostics)

        /**
         * return null if Audit fails
         */
        if (!Object.prototype.hasOwnProperty.call(result, 'details')) {
            return null
        }

        return result.details.items[0]
    }

    async getMetrics () {
        const serverResponseTime = await this._audit(ServerResponseTime, { URL: this.url })
        const cumulativeLayoutShift = await this._audit(CumulativeLayoutShift)
        const result = await this._audit(Metrics)
        const metrics = result.details.items[0] || {}
        return {
            estimatedInputLatency: metrics.estimatedInputLatency,
            /**
             * keeping TTFB for backwards compatibility
             */
            timeToFirstByte: Math.round(serverResponseTime.numericValue, 10),
            serverResponseTime: Math.round(serverResponseTime.numericValue, 10),
            domContentLoaded: metrics.observedDomContentLoaded,
            firstVisualChange: metrics.observedFirstVisualChange,
            firstPaint: metrics.observedFirstPaint,
            firstContentfulPaint: metrics.firstContentfulPaint,
            firstMeaningfulPaint: metrics.firstMeaningfulPaint,
            largestContentfulPaint: metrics.largestContentfulPaint,
            lastVisualChange: metrics.observedLastVisualChange,
            firstCPUIdle: metrics.firstCPUIdle,
            firstInteractive: metrics.interactive,
            load: metrics.observedLoad,
            speedIndex: metrics.speedIndex,
            totalBlockingTime: metrics.totalBlockingTime,
            cumulativeLayoutShift: cumulativeLayoutShift.numericValue,
        }
    }

    async getPerformanceScore() {
        const auditResults = {}
        auditResults['speed-index'] =  await this._audit(SpeedIndex)
        auditResults['first-contentful-paint'] =  await this._audit(FirstContentfulPaint)
        auditResults['largest-contentful-paint'] =  await this._audit(LargestContentfulPaint)
        auditResults['cumulative-layout-shift'] =  await this._audit(CumulativeLayoutShift)
        auditResults['total-blocking-time'] =  await this._audit(TotalBlockingTime)
        auditResults.interactive = await this._audit(InteractiveMetric)

        if (!auditResults.interactive || !auditResults['cumulative-layout-shift'] || !auditResults['first-contentful-paint'] ||
            !auditResults['largest-contentful-paint'] || !auditResults['speed-index'] || !auditResults['total-blocking-time']) {
            log.info('One or multiple required metrics couldn\'t be found, setting performance score to: null')
            return null
        }

        const scores = defaultConfig.categories.performance.auditRefs.filter((auditRef) => auditRef.weight).map((auditRef) => ({
            score: auditResults[auditRef.id].score,
            weight: auditRef.weight,
        }))
        return ReportScoring.arithmeticMean(scores)
    }
}
