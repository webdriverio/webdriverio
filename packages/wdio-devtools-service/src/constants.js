/**
 * performance tracing categories
 */
export const DEFAULT_TRACING_CATEGORIES = [
    '-*', 'devtools.timeline', 'v8.execute', 'disabled-by-default-devtools.timeline',
    'disabled-by-default-devtools.timeline.frame', 'toplevel',
    'blink.console', 'blink.user_timing', 'latencyInfo', 'disabled-by-default-devtools.timeline.stack',
    'disabled-by-default-v8.cpu_profiler', 'disabled-by-default-devtools.screenshot'
]
