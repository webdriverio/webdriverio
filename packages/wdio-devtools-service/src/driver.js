import puppeteer from 'puppeteer-core'

import logger from '@wdio/logger'

const log = logger('@wdio/devtools-service:DevToolsDriver')

/**
 * connects a puppeteer instance to the browser
 */
export default class DevToolsDriver {
    constructor (browser) {
        this.browser = browser

        /**
         * ToDo (Christian): handle target changes
         * this should usually never happens as we only connect within one
         * target within the VM so maybe we can discard this
         */
    }

    get devices () {
        return puppeteer.devices
    }

    static async attach (url) {
        log.info(`Connect to ${url}`)
        const browser = await puppeteer.connect({
            browserURL: url,
            defaultViewport: null
        })
        return new DevToolsDriver(browser)
    }

    async getActiveTarget () {
        if (this.target) {
            return this.target
        }

        const target = this.target = await this.browser.waitForTarget(
            /* istanbul ignore next */
            (t) => t.type() === 'page')
        return target
    }

    async getActivePage () {
        if (this.page) {
            return this.page
        }

        const target = await this.getActiveTarget()
        this.page = await target.page()
        return this.page
    }

    async getCDPSession () {
        if (this.cdpSession) {
            return this.cdpSession
        }

        const target = await this.getActiveTarget()
        this.cdpSession = await target.createCDPSession()
        return this.cdpSession
    }

    async send (method, params) {
        const cdpSession = await this.getCDPSession()
        return cdpSession.send(method, params)
    }
}
