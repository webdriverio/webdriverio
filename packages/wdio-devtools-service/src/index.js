import fs from 'fs'
import path from 'path'
import logger from 'wdio-logger'
import CDP from 'chrome-remote-interface'

const log = logger('wdio-devtools-service')

const UNSUPPORTED_ERROR_MESSAGE = 'The wdio-devtools-service currently only supports Chrome version 63 and up'
const RE_DEVTOOLS_DEBUGGING_PORT_SWITCH = /--remote-debugging-port=(\d*)/
const RE_USER_DATA_DIR_SWITCH = /--user-data-dir=([^-]*)/

export default class DevToolsService {
    constructor () {
        this.isSupported = false
    }

    beforeSession (_, caps) {
        if (caps.browserName !== 'chrome' || (caps.version && caps.version < 63)) {
            return log.error(UNSUPPORTED_ERROR_MESSAGE)
        }

        this.isSupported = true
    }

    async before () {
        if (!this.isSupported) {
            return global.browser.addCommand('cdp', /* istanbul ignore next */ () => {
                throw new Error(UNSUPPORTED_ERROR_MESSAGE)
            })
        }

        try {
            this.chromePort = await this._findChromePort()
        } catch (err) {
            log.error(`Couldn't connect to chrome: ${err.stack}`)
            return
        }

        this.client = await this._getCDPClient(this.chromePort)

        /**
         * allow to easily access the CDP from the browser object
         */
        global.browser.addCommand('cdp', (domain, command, args = {}) => {
            if (!this.client[domain]) {
                throw new Error(`Domain "${domain}" doesn't exist in the Chrome DevTools protocol`)
            }

            if (!this.client[domain][command]) {
                throw new Error(`The "${domain}" domain doesn't have a method called "${command}"`)
            }

            log.info(`Send command "${domain}.${command}" with args: ${JSON.stringify(args)}`)
            return new Promise((resolve, reject) => this.client[domain][command](args, (err, result) => {
                /* istanbul ignore if */
                if (err) {
                    return reject(new Error(`Chrome DevTools Error: ${result.message}`))
                }

                return resolve(result)
            }))
        })

        /**
         * helper method to receive Chrome remote debugging connection data to
         * e.g. use external tools like lighthouse
         */
        const { host, port } = this.client
        global.browser.addCommand('cdpConnection', () => ({ host, port }))

        /**
         * propagate CDP events to the browser event listener
         */
        this.client.on('event', (event) => {
            const method = event.method || 'event'
            log.debug(`cdp event: ${method} with params ${JSON.stringify(event.params)}`)
            global.browser.emit(method, event.params)
        })
    }

    /**
     * Find Chrome DevTools Interface port by checking Chrome switches from the chrome://version
     * page. In case a newer version is used (+v65) we check the DevToolsActivePort file
     */
    async _findChromePort () {
        await global.browser.url('chrome://version')
        const cmdLineTextElem = await global.browser.$('#command_line')
        const cmdLineText = await cmdLineTextElem.getText()
        let port = parseInt(cmdLineText.match(RE_DEVTOOLS_DEBUGGING_PORT_SWITCH)[1], 10)

        /**
         * newer Chrome versions store port in DevToolsActivePort file
         */
        if (port === 0) {
            const userDataDir = cmdLineText.match(RE_USER_DATA_DIR_SWITCH)[1].trim()
            const devToolsActivePortFile = fs.readFileSync(path.join(userDataDir, 'DevToolsActivePort'), 'utf8')
            port = devToolsActivePortFile.split('\n').shift()
        }

        return port
    }

    async _getCDPClient (port) {
        return new Promise((resolve) => CDP({
            port,
            host: 'localhost',
            target: /* istanbul ignore next */ (targets) => targets.findIndex((t) => t.type === 'page')
        }, resolve))
    }
}
