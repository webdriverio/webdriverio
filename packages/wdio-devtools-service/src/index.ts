import logger from '@wdio/logger'
import puppeteerCore from 'puppeteer-core'

import type { Page } from 'puppeteer-core/lib/cjs/puppeteer/common/Page'
import type { CDPSession } from 'puppeteer-core/lib/cjs/puppeteer/common/Connection'
import type { Viewport } from 'puppeteer-core/lib/cjs/puppeteer/common/PuppeteerViewport'
import type { Browser } from 'puppeteer-core/lib/cjs/puppeteer/common/Browser'
import type { Target } from 'puppeteer-core/lib/cjs/puppeteer/common/Target'

import CommandHandler from './commands'
import Auditor from './auditor'
import TraceGatherer from './gatherer/trace'
import DevtoolsGatherer, { CDPSessionOnMessageObject } from './gatherer/devtools'
import { isBrowserSupported, setUnsupportedCommand } from './utils'
import { NETWORK_STATES, UNSUPPORTED_ERROR_MESSAGE, CLICK_TRANSITION, DEFAULT_THROTTLE_STATE } from './constants'

const log = logger('@wdio/devtools-service')
const TRACE_COMMANDS = ['click', 'navigateTo', 'url']

interface EnablePerformanceAuditsOptions {
    cacheEnabled: boolean
    cpuThrottling: number
    networkThrottling: keyof typeof NETWORK_STATES
}

interface DeviceDescription {
    viewport: Viewport;
    userAgent: string;
}

interface Device {
    name: string;
    userAgent: string;
    viewport: {
        width: number;
        height: number;
        deviceScaleFactor: number;
        isMobile: boolean;
        hasTouch: boolean;
        isLandscape: boolean;
    };
}

export default class DevToolsService implements WebdriverIO.ServiceInstance {
    private _isSupported = false
    private _shouldRunPerformanceAudits = false

    private _puppeteer?: Browser
    private _target?: Target
    private _page: Page | null = null
    private _session?: CDPSession
    private _cacheEnabled?: boolean
    private _cpuThrottling?: number
    private _networkThrottling?: keyof typeof NETWORK_STATES
    private _traceGatherer?: TraceGatherer
    private _devtoolsGatherer?: DevtoolsGatherer
    private _browser?: WebdriverIO.BrowserObject | WebdriverIO.MultiRemoteBrowserObject

    beforeSession (_: WebdriverIO.Config, caps: WebDriver.DesiredCapabilities) {
        if (!isBrowserSupported(caps)) {
            return log.error(UNSUPPORTED_ERROR_MESSAGE)
        }
        this._isSupported = true
    }

    before (
        caps: WebDriver.Capabilities,
        specs: string[],
        browser: WebdriverIO.BrowserObject | WebdriverIO.MultiRemoteBrowserObject
    ) {
        this._browser = browser
        this._isSupported = this._isSupported || Boolean(this._browser.puppeteer)
        return this._setupHandler()
    }

    async onReload () {
        if (!this._browser) {
            return
        }

        // resetting puppeteer on sessionReload, so a new puppeteer session will be attached
        this._browser.puppeteer = null
        return this._setupHandler()
    }

    async beforeCommand (commandName: string, params: any[]) {
        if (!this._shouldRunPerformanceAudits || !this._traceGatherer || this._traceGatherer.isTracing || !TRACE_COMMANDS.includes(commandName)) {
            return
        }

        /**
         * set browser profile
         */
        this._setThrottlingProfile(this._networkThrottling, this._cpuThrottling, this._cacheEnabled)

        const url = ['url', 'navigateTo'].some(cmdName => cmdName === commandName)
            ? params[0]
            : CLICK_TRANSITION
        return this._traceGatherer.startTracing(url)
    }

    async afterCommand (commandName: string) {
        if (!this._traceGatherer || !this._traceGatherer.isTracing || !TRACE_COMMANDS.includes(commandName)) {
            return
        }

        /**
         * update custom commands once tracing finishes
         */
        this._traceGatherer.once('tracingComplete', (traceEvents) => {
            const auditor = new Auditor(traceEvents, this._devtoolsGatherer?.getLogs())
            auditor.updateCommands(this._browser as WebdriverIO.BrowserObject)
        })

        this._traceGatherer.once('tracingError', (err: Error) => {
            const auditor = new Auditor()
            auditor.updateCommands(this._browser as WebdriverIO.BrowserObject, /* istanbul ignore next */() => {
                throw new Error(`Couldn't capture performance due to: ${err.message}`)
            })
        })

        return new Promise<void>((resolve) => {
            log.info(`Wait until tracing for command ${commandName} finishes`)

            /**
             * wait until tracing stops
             */
            this._traceGatherer?.once('tracingFinished', async () => {
                log.info('Disable throttling')
                await this._setThrottlingProfile('online', 0, true)

                log.info('continuing with next WebDriver command')
                resolve()
            })
        })
    }

    /**
     * set flag to run performance audits for page transitions
     */
    _enablePerformanceAudits ({ networkThrottling, cpuThrottling, cacheEnabled }: EnablePerformanceAuditsOptions = DEFAULT_THROTTLE_STATE) {
        if (!NETWORK_STATES[networkThrottling]) {
            throw new Error(`Network throttling profile "${networkThrottling}" is unknown, choose between ${Object.keys(NETWORK_STATES).join(', ')}`)
        }

        if (typeof cpuThrottling !== 'number') {
            throw new Error(`CPU throttling rate needs to be typeof number but was "${typeof cpuThrottling}"`)
        }

        this._networkThrottling = networkThrottling
        this._cpuThrottling = cpuThrottling
        this._cacheEnabled = Boolean(cacheEnabled)
        this._shouldRunPerformanceAudits = true
    }

    /**
     * custom command to disable performance audits
     */
    _disablePerformanceAudits () {
        this._shouldRunPerformanceAudits = false
    }

    /**
     * set device emulation
     */
    async _emulateDevice (device: string | DeviceDescription, inLandscape?: boolean) {
        if (!this._page) {
            throw new Error('No page has been captured yet')
        }

        if (typeof device === 'string') {
            const deviceName = device + (inLandscape ? ' landscape' : '')
            const deviceCapabilities = puppeteerCore.devices[deviceName]
            if (!deviceCapabilities) {
                const deviceNames = (puppeteerCore.devices as any)
                    .map((device: Device) => device.name)
                    .filter((device: string) => !device.endsWith('landscape'))
                throw new Error(`Unknown device, available options: ${deviceNames.join(', ')}`)
            }

            return this._page.emulate(deviceCapabilities)
        }

        return this._page.emulate(device)
    }

    /**
     * helper method to set throttling profile
     */
    async _setThrottlingProfile(
        networkThrottling = DEFAULT_THROTTLE_STATE.networkThrottling,
        cpuThrottling = DEFAULT_THROTTLE_STATE.cpuThrottling,
        cacheEnabled = DEFAULT_THROTTLE_STATE.cacheEnabled
    ) {
        if (!this._page || !this._session) {
            throw new Error('No page or session has been captured yet')
        }

        await this._page.setCacheEnabled(Boolean(cacheEnabled))
        await this._session.send('Emulation.setCPUThrottlingRate', { rate: cpuThrottling })
        await this._session.send('Network.emulateNetworkConditions', NETWORK_STATES[networkThrottling])
    }

    async _setupHandler () {
        if (!this._isSupported || !this._browser) {
            return setUnsupportedCommand(this._browser as WebdriverIO.BrowserObject)
        }

        this._puppeteer = await this._browser.getPuppeteer() as unknown as Browser
        /* istanbul ignore next */
        if (!this._puppeteer) {
            throw new Error('Could not initiate Puppeteer instance')
        }

        this._target = await this._puppeteer.waitForTarget(
            /* istanbul ignore next */
            (t) => t.type() === 'page')
        /* istanbul ignore next */
        if (!this._target) {
            throw new Error('No page target found')
        }

        this._page = await this._target.page()
        /* istanbul ignore next */
        if (!this._page) {
            throw new Error('No page found')
        }

        this._session = await this._target.createCDPSession()

        new CommandHandler(this._session, this._page, this._browser)
        this._traceGatherer = new TraceGatherer(this._session, this._page)

        this._session.on('Page.loadEventFired', this._traceGatherer.onLoadEventFired.bind(this._traceGatherer))
        this._session.on('Page.frameNavigated', this._traceGatherer.onFrameNavigated.bind(this._traceGatherer))

        this._page.on('requestfailed', this._traceGatherer.onFrameLoadFail.bind(this._traceGatherer))

        /**
         * enable domains for client
         */
        await Promise.all(['Page', 'Network', 'Console'].map(
            (domain) => Promise.all([
                this._session?.send(`${domain}.enable` as any)
            ])
        ))

        this._devtoolsGatherer = new DevtoolsGatherer()
        this._puppeteer['_connection']._transport._ws.addEventListener('message', (event: { data: string }) => {
            const data: CDPSessionOnMessageObject = JSON.parse(event.data)
            this._devtoolsGatherer?.onMessage(data)
            const method = data.method || 'event'
            log.debug(`cdp event: ${method} with params ${JSON.stringify(data.params)}`)

            if (this._browser) {
                this._browser.emit(method, data.params)
            }
        })

        this._browser.addCommand('enablePerformanceAudits', this._enablePerformanceAudits.bind(this))
        this._browser.addCommand('disablePerformanceAudits', this._disablePerformanceAudits.bind(this))
        this._browser.addCommand('emulateDevice', this._emulateDevice.bind(this))
    }
}
