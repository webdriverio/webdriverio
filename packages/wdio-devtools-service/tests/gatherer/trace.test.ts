import type Protocol from 'devtools-protocol'
import type { CDPSession } from 'puppeteer-core/lib/cjs/puppeteer/common/Connection'
import type { Page } from 'puppeteer-core/lib/cjs/puppeteer/common/Page'
import type { HTTPRequest } from 'puppeteer-core/lib/cjs/puppeteer/common/HTTPRequest'
import type { WaitPromise } from '../../src/gatherer/trace'

import TraceGatherer from '../../src/gatherer/trace'
import { FRAME_LOAD_START_TIMEOUT, CLICK_TRANSITION } from '../../src/constants'

import TRACELOG from '../__fixtures__/tracelog.json'

let traceGatherer: TraceGatherer

const pageMock = {
    on: jest.fn(),
    tracing: {
        start: jest.fn(),
        stop: jest.fn().mockReturnValue(Promise.resolve(Buffer.from(JSON.stringify(TRACELOG))))
    },
    evaluateOnNewDocument: jest.fn()
}

const sessionMock = {
    on: jest.fn(),
    emit: jest.fn(),
    send: jest.fn(),
    removeListener: jest.fn()
}

const frame = {
    id: '123',
    loaderId: 'foobar123',
    url: 'http://foobar.com'
}

jest.useFakeTimers()

beforeEach(() => {
    pageMock.on.mockClear()
    pageMock.tracing.start.mockClear()
    pageMock.tracing.stop.mockClear()
    pageMock.evaluateOnNewDocument.mockClear()
    sessionMock.on.mockClear()
    traceGatherer = new TraceGatherer(
        sessionMock as unknown as CDPSession,
        pageMock as unknown as Page
    )
    traceGatherer.emit = jest.fn()
})

test('should register eventlisteners for network monitor', () => {
    traceGatherer['_networkStatusMonitor'] = { dispatch: jest.fn() }
    for (const fn of Object.values(traceGatherer['_networkListeners'])) {
        fn({ some: 'params' })
    }
    expect(traceGatherer['_networkStatusMonitor'].dispatch.mock.calls)
        .toMatchSnapshot()
})

test('onFrameLoadFail', () => {
    traceGatherer.onFrameLoadFail({
        frame: () => undefined
    } as unknown as HTTPRequest)
    expect(traceGatherer['_failingFrameLoadIds']).toEqual([])
    traceGatherer.onFrameLoadFail({
        frame: () => ({ _id: '123' })
    } as unknown as HTTPRequest)
    expect(traceGatherer['_failingFrameLoadIds']).toEqual(['123'])
})

test('getter isTracing', () => {
    expect(traceGatherer.isTracing).toBe(false)

    traceGatherer['_traceStart'] = Date.now()
    expect(traceGatherer.isTracing).toBe(true)
})

test('finishTracing', async () => {
    traceGatherer['_pageLoadDetected'] = true
    traceGatherer['_networkStatusMonitor'] = true
    traceGatherer['_traceStart'] = 123
    traceGatherer['_frameId'] = 'true'
    traceGatherer['_loaderId'] = 'true'
    traceGatherer['_pageUrl'] = 'http://some.url'
    traceGatherer['_waitForNetworkIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise
    traceGatherer['_waitForCPUIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise

    traceGatherer.finishTracing()
    await new Promise((resolve) => process.nextTick(resolve))

    expect(traceGatherer['_pageLoadDetected']).toBe(false)
    expect(traceGatherer['_networkStatusMonitor']).toBe(undefined)
    expect(traceGatherer['_traceStart']).toBe(undefined)
    expect(traceGatherer['_frameId']).toBe(undefined)
    expect(traceGatherer['_loaderId']).toBe(undefined)
    expect(traceGatherer['_pageUrl']).toBe(undefined)

    expect(sessionMock.removeListener).toHaveBeenCalledTimes(7)
    expect(traceGatherer['_waitForNetworkIdleEvent'].cancel).toHaveBeenCalledTimes(1)
    expect(traceGatherer['_waitForCPUIdleEvent'].cancel).toHaveBeenCalledTimes(1)
    expect(traceGatherer.emit).toBeCalledWith('tracingFinished')
})

test('onFrameNavigated', () => {
    expect(traceGatherer['_frameId']).toBe(undefined)
    expect(traceGatherer['_loaderId']).toBe(undefined)
    expect(traceGatherer['_pageUrl']).toBe(undefined)
    traceGatherer['_traceStart'] = Date.now()

    traceGatherer.onFrameNavigated({ frame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toBeCalledWith('tracingStarted', '123')

    expect(traceGatherer['_frameId']).toBe('123')
    expect(traceGatherer['_loaderId']).toBe('foobar123')
    expect(traceGatherer['_pageUrl']).toBe('http://foobar.com')
})

test('onFrameNavigated: should not start if frame was detected', () => {
    traceGatherer['_frameId'] = '123'
    traceGatherer.onFrameNavigated({ frame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toHaveBeenCalledTimes(0)
})

test('onFrameNavigated: should not start if a subframe was detected', () => {
    const subFrame = Object.assign({}, frame, { parentId: 122 })
    traceGatherer.onFrameNavigated({ frame: subFrame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toHaveBeenCalledTimes(0)
})

test('onFrameNavigated: should not start if url is not supported', () => {
    const subFrame = Object.assign({}, frame, { url: 'data:,foobar' })
    traceGatherer.onFrameNavigated({ frame: subFrame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toHaveBeenCalledTimes(0)
})

test('onFrameNavigated: should not start if tracing is not started', () => {
    traceGatherer['_traceStart'] = undefined
    traceGatherer.finishTracing = jest.fn()
    traceGatherer['_failingFrameLoadIds'].push('123')
    traceGatherer['_waitForNetworkIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise
    traceGatherer['_waitForCPUIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise
    traceGatherer.onFrameNavigated({ frame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toHaveBeenCalledTimes(0)
    expect(traceGatherer.finishTracing).toHaveBeenCalledTimes(0)
    expect(traceGatherer['_waitForNetworkIdleEvent'].cancel).toHaveBeenCalledTimes(0)
    expect(traceGatherer['_waitForCPUIdleEvent'].cancel).toHaveBeenCalledTimes(0)
})

test('onFrameNavigated: should cancel trace if page load failed', () => {
    traceGatherer['_traceStart'] = Date.now()
    traceGatherer.finishTracing = jest.fn()
    traceGatherer['_failingFrameLoadIds'].push('123')
    traceGatherer['_waitForNetworkIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise
    traceGatherer['_waitForCPUIdleEvent'] = { cancel: jest.fn() } as unknown as WaitPromise
    traceGatherer.onFrameNavigated({ frame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer.emit).toHaveBeenCalledWith('tracingError', expect.any(Error))
    expect(traceGatherer.finishTracing).toHaveBeenCalledTimes(1)
    expect(traceGatherer['_waitForNetworkIdleEvent'].cancel).toHaveBeenCalledTimes(1)
    expect(traceGatherer['_waitForCPUIdleEvent'].cancel).toHaveBeenCalledTimes(1)
})

test('onFrameNavigated: can detect page load', () => {
    traceGatherer['_clickTraceTimeout'] = 123 as unknown as NodeJS.Timeout
    traceGatherer['_traceStart'] = Date.now()

    expect(traceGatherer['_pageLoadDetected']).toBe(false)
    traceGatherer.onFrameNavigated({ frame } as unknown as Protocol.Page.FrameNavigatedEvent)
    expect(traceGatherer['_pageLoadDetected']).toBe(true)
})

test('startTracing', async () => {
    traceGatherer.waitForNetworkIdle = jest.fn()
    traceGatherer.waitForCPUIdle = jest.fn()
    await traceGatherer.startTracing('http://foobar.com')

    expect(sessionMock.on).toHaveBeenCalledTimes(7)
    expect(pageMock.tracing.start).toHaveBeenCalledTimes(1)
    expect(pageMock.evaluateOnNewDocument).toHaveBeenCalledTimes(1)
    expect(traceGatherer.waitForNetworkIdle).toHaveBeenCalledTimes(1)
    expect(traceGatherer.waitForCPUIdle).toHaveBeenCalledTimes(1)
})

test('startTracing: registers timeout for click events', async () => {
    traceGatherer.waitForNetworkIdle = jest.fn()
    traceGatherer.waitForCPUIdle = jest.fn()
    traceGatherer.finishTracing = jest.fn()

    await traceGatherer.startTracing(CLICK_TRANSITION)
    jest.advanceTimersByTime(FRAME_LOAD_START_TIMEOUT + 10)
    expect(pageMock.tracing.stop).toHaveBeenCalledTimes(1)
    await new Promise((resolve) => process.nextTick(resolve))
    expect(traceGatherer.finishTracing).toHaveBeenCalledTimes(1)
})

test('completeTracing', async () => {
    traceGatherer.finishTracing = jest.fn()
    await traceGatherer.completeTracing()
    expect(traceGatherer.finishTracing).toHaveBeenCalledTimes(1)
    expect(traceGatherer.emit).toBeCalledWith('tracingComplete', expect.any(Object))
})

test('completeTracing: in failure case', async () => {
    traceGatherer.finishTracing = jest.fn()
    pageMock.tracing.stop.mockReturnValue(Promise.reject(new Error('boom')))
    await traceGatherer.completeTracing()
    expect(traceGatherer.finishTracing).toHaveBeenCalledTimes(1)
    expect(traceGatherer.emit).toBeCalledWith('tracingError', expect.any(Error))
})

test('onLoadEventFired', async () => {
    await traceGatherer.onLoadEventFired()
    traceGatherer['_traceStart'] = Date.now()
    traceGatherer.completeTracing = jest.fn().mockReturnValue(Promise.resolve('yeahh'))

    traceGatherer['_waitForNetworkIdleEvent'] = {
        promise: new Promise((resolve) => {
            jest.advanceTimersByTime(50)
            resolve()
        }),
        cancel: jest.fn()
    }
    traceGatherer['_waitForCPUIdleEvent'] = {
        promise: new Promise((resolve) => {
            jest.advanceTimersByTime(100)
            resolve()
        }),
        cancel: jest.fn()
    }

    process.nextTick(() => jest.advanceTimersByTime(10000))
    await traceGatherer.onLoadEventFired()
    expect(traceGatherer['_waitForNetworkIdleEvent'].cancel).toHaveBeenCalledTimes(1)
    expect(traceGatherer['_waitForCPUIdleEvent'].cancel).toHaveBeenCalledTimes(1)
    expect(traceGatherer.completeTracing).toHaveBeenCalledTimes(1)
})

test('onLoadEventFired: using min trace time', async () => {
    traceGatherer['_traceStart'] = Date.now() - 9700
    traceGatherer.completeTracing = jest.fn().mockReturnValue(Promise.resolve('yeahh'))

    const doneCb = jest.fn()
    traceGatherer['_waitForNetworkIdleEvent'] = {
        promise: new Promise((resolve) => {
            jest.advanceTimersByTime(50)
            resolve()
        }),
        cancel: jest.fn()
    }
    traceGatherer['_waitForCPUIdleEvent'] = {
        promise: new Promise((resolve) => {
            jest.advanceTimersByTime(10)
            resolve()
        }),
        cancel: jest.fn()
    }
    traceGatherer.waitForMaxTimeout = jest.fn().mockReturnValue(new Promise((resolve) => {
        jest.advanceTimersByTime(150)
        resolve(doneCb)
    }))

    await traceGatherer.onLoadEventFired()
    expect(doneCb).toBeCalledTimes(1)
})

test('waitForMaxTimeout', async () => {
    traceGatherer.completeTracing = jest.fn()
    process.nextTick(() => jest.advanceTimersByTime(200))
    const done = await traceGatherer.waitForMaxTimeout(200)
    done()
    expect(traceGatherer.completeTracing).toHaveBeenCalledTimes(1)
})
