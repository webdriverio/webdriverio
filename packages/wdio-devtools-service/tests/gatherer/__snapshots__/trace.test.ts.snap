// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should register eventlisteners for network monitor 1`] = `
Array [
  Array [
    Object {
      "method": "Network.requestWillBeSent",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.requestServedFromCache",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.responseReceived",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.dataReceived",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.loadingFinished",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.loadingFailed",
      "params": Object {
        "some": "params",
      },
    },
  ],
  Array [
    Object {
      "method": "Network.resourceChangedPriority",
      "params": Object {
        "some": "params",
      },
    },
  ],
]
`;
